receiver/oracledbreceiver/scraper.go

```diff
@@ -0,0 +1,306 @@
+	sessionExchangeDeadlocksSQL   = "select ss.SID as session_id, se.value as VALUE from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and se.SID = ss.SID and NAME = 'exchange deadlocks'"
+	sessionExecuteCountSQL        = "select ss.SID as session_id, se.value as VALUE  from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and se.SID = ss.SID and NAME = 'execute count'"
+	sessionParseCountTotalSQL     = "select ss.SID as session_id, se.value as VALUE from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and se.SID = ss.SID and NAME = 'parse count (total)'"
+	sessionUserCommitsSQL         = "select ss.SID as session_id, se.value as VALUE from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and se.SID = ss.SID and NAME = 'user commits'"
+	sessionUserRollbacksSQL       = "select ss.SID as session_id, se.value as VALUE from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and se.SID = ss.SID and NAME = 'user rollbacks'"
+	activeSessionTotalSQL         = "select status, type, count(*) as VALUE FROM v$session GROUP BY status, type"
+	cachedSessionTotalSQL         = "select type, count(*) as VALUE FROM v$session WHERE status = 'CACHED' GROUP BY type"
+)
+
+type dbProviderFunc func() (*sql.DB, error)
+
+type clientProviderFunc func(*sql.DB, string, *zap.Logger) dbClient
+
+type scraper struct {
+	logger             *zap.Logger
+	metricsBuilder     *metadata.MetricsBuilder
+	dbProviderFunc     dbProviderFunc
+	clientProviderFunc clientProviderFunc
+	db                 *sql.DB
+	id                 config.ComponentID
+	scrapeCfg          scraperhelper.ScraperControllerSettings
+	startTime          pcommon.Timestamp
+	metricsSettings    metadata.MetricsSettings
+}
+
+func newScraper(id config.ComponentID, metricsBuilder *metadata.MetricsBuilder, metricsSettings metadata.MetricsSettings, scrapeCfg scraperhelper.ScraperControllerSettings, logger *zap.Logger, providerFunc dbProviderFunc, clientProviderFunc clientProviderFunc) *scraper {
+	return &scraper{
+		id:                 id,
+		metricsBuilder:     metricsBuilder,
+		metricsSettings:    metricsSettings,
+		scrapeCfg:          scrapeCfg,
+		logger:             logger,
+		dbProviderFunc:     providerFunc,
+		clientProviderFunc: clientProviderFunc,
+	}
+}
+
+var _ scraperhelper.Scraper = (*scraper)(nil)
+
+func (s *scraper) ID() config.ComponentID {
+	return s.id
+}
+
+func (s *scraper) Start(context.Context, component.Host) error {
+	s.startTime = pcommon.NewTimestampFromTime(time.Now())
+	var err error
+	s.db, err = s.dbProviderFunc()
+	if err != nil {
+		return fmt.Errorf("failed to open db connection: %w", err)
+	}
+	return nil
+}
+
+func (s *scraper) Scrape(ctx context.Context) (pmetric.Metrics, error) {
+	s.logger.Debug("Begin scrape")
+	now := pcommon.NewTimestampFromTime(time.Now())
```

Comment:
Using a static ts and unreliable query performance could introduce some unintuitive drift where updated w/ the query would be more accurate.

Reply:
I wasn't sure about this. I used the approach taken by sqlqueryreceiver, but I had the same gut reaction as you have here.
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/sqlqueryreceiver/scraper.go#L65

I'd feel comfortable changing to your suggestion.

Reply:
unless you want to keep it transparent, addressing this in your sqlquery lib PR could make sense

Reply:
Good call, I'll do it in both places.